### DISCUSSION TITLE
Forwarding downstream PLIs from readers back up to publishers

### DISCUSSION THREAD

**Discussion by tmatth** [bluenviron/mediamtx#3567] [Questions]
### Question

I'm using `mediamtx` as a proxy to another RTSP server. Readers reading from `mediamtx` will send a PLI when needed, but I'd like to forward that PLI to the upstream source/publisher. I believe `mediamtx` simply swallows the incoming RTCP traffic it gets from players. I've added some logging here:

```diff
+diff --git a/internal/servers/rtsp/session.go b/internal/servers/rtsp/session.go
+index a3e7a90..eab1bb9 100644
+--- a/internal/servers/rtsp/session.go
++++ b/internal/servers/rtsp/session.go
+@@ -11,8 +11,10 @@ import (
+       "github.com/bluenviron/gortsplib/v4"
+       rtspauth "github.com/bluenviron/gortsplib/v4/pkg/auth"
+       "github.com/bluenviron/gortsplib/v4/pkg/base"
++      "github.com/bluenviron/gortsplib/v4/pkg/description"
+       "github.com/google/uuid"
+       "github.com/pion/rtp"
++      "github.com/pion/rtcp"
+ 
+       "github.com/bluenviron/mediamtx/internal/auth"
+       "github.com/bluenviron/mediamtx/internal/conf"
+@@ -259,6 +261,19 @@ func (s *session) onPlay(_ *gortsplib.ServerHandlerOnPlayCtx) (*base.Response, e
+                       Query:           s.rsession.SetuppedQuery(),
+               })
+ 
++              for _, medi := range s.rsession.SetuppedMedias() {
++                      for _, _ = range medi.Formats {
++                              s.rsession.OnPacketRTCPAny(func(medi *description.Media, pkt rtcp.Packet) {
++                                      switch p := pkt.(type) {
++                                      case *rtcp.PictureLossIndication:
++                                              s.Log(logger.Warn, "GOT PLI packet %v, %v", pkt, p)
++                                              // TODO: forward this PLI to original publisher
++                                      default:
++                                              s.Log(logger.Warn, "GOT RTCP packet %v, %v", pkt, p)
++                                      }
++                              })
++                      }
++              }
++
+               s.mutex.Lock()
+               s.state = gortsplib.ServerSessionStatePlay
+               s.transport = s.rsession.SetuppedTransport()
```
How can I access the underlying publisher and forward the RTCP PLI packet to it?

**Comment by neilyoung** [comment]
That would interest me too...

**Comment by aler9** [comment]
Hello, forwarding RTCP reports or PLIs from every reader to the source is not a good idea, since:

* the behavior of a single laggy reader would influence every other reader
* the source would receive N PLIs instead of 1 PLI (where N is the number of readers), with the risk of multiple IDR frames being generated in a few instants.

The connection between the server and the source is completely isolated from the connection between the server and each reader and it will remain in this way. Eventually, we can improve the server in order to periodically send PLIs with RTSP. This is already performed with WebRTC.

**Reply by tmatth** [reply]
I agree you wouldn't want to blindly spam a source with RTCP traffic if you have no means of filtering and/or throttling it, however in my use case I'm only interested in PLIs for now and specifically in reducing the time it takes between a new reader joining and being able to play back the stream, so really the PLI(s) emitted when a reader starts.

Without any downstream PLI information, the obvious approach is emitting PLIs at a fixed interval like you mention. The downside is that you'll end up sending a lot more keyframes than necessary (say you have a bunch of readers join at the beginning and then watch for an hour without any connection issues, they don't need PLIs every N seconds). The other downside is that every new viewer will have up to wait up to N seconds before the stream will play.

Moreover, since my source (that is being proxied by mediamtx) is the Janus videoroom plugin it actually has builtin throttling for PLIs so there's no risk of spamming the true upstream source (i.e. the browser publishing to the videoroom plugin).

In any case, for now I've added some questionable hacks to both Janus and mediamtx to relay PLIs (and map them to the correct upstream publisher) but I'd love for there to be a native way of achieving this.

**Reply by aler9** [reply]
PLIs are part of the AVPF / SAVPF profile, which is now planned and tracked here: https://github.com/bluenviron/gortsplib/issues/901

So in the future we'll have them.
